[FLINK-28229][streaming-java] Introduce Source API alternatives for StreamExecutionEnvironment#fromCollection() methods

tigrulya-exe opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

This pull request adds FLIP-27 Source API alternatives for `FromElementsFunction` and `FromIteratorFunction`.

## Brief change log

  - Added CollectionSource
  - Added IteratorSource

## Verifying this change

This change added tests and can be verified as follows:

  - Added tests for both sources, validating elements emission, snapshot and recovery from snapshot logic.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (yes)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

zentol commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/connector/source/IteratorSource.java:
```diff
@@ -0,0 +1,137 @@
+
+    @Override
+    public Boundedness getBoundedness() {
+        return boundedness;
```
We can assume this to be bounded imo.

tigrulya-exe commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/connector/source/CollectionSource.java:
```diff
@@ -0,0 +1,219 @@
+    @Override
+    public void setOutputType(TypeInformation<E> outTypeInfo, ExecutionConfig executionConfig) {
+        Preconditions.checkState(
+                elements != null,
```
Replaced with `outTypeInfo != null` check

tigrulya-exe commented:
Hi, @zentol ! I've resolved comments, replaced old sources in related methods of StreamExEnv with new ones and fixed all failed tests that followed this replacement. Now I'm waiting for your decision about `DataGeneratorSource`

afedulov commented:
@zentol One concerns regarding the idea to migrate the existing methods in `StreamExecutionEnvironment` to `DataGeneratorSource` is that adding the `flink-streaming-java` dependency on `flink-connector-datagen` introduces a somewhat unexpected cycle between `flink-architecture-tests-test` and `flink-clients`. How can we address this?

afedulov commented:
@zentol I implemented the alternative approach you proposed based on the `DataGeneratorSource`. There are a couple of open questions, but I'd like to get your general feedback before working on them.
